candle-transformers/src/models/stable_diffusion/uni_pc.rs

```diff
@@ -0,0 +1,1048 @@
+        }
+
+        pub fn add(&mut self, x: f64) {
+            self.max = self.max.map(|y| y.max(x)).or(Some(x));
+
+            if self.n[4] < 5 {
+                self.q[self.n[4] as usize] = x;
+                self.n[4] += 1;
+                if self.n[4] == 5 {
+                    sort_floats(&mut self.q);
+                }
+                return;
+            }
+
+            let mut k: usize;
+            if x < self.q[0] {
+                self.q[0] = x;
+                k = 0;
+            } else {
+                k = 4;
+                for i in 1..5 {
+                    if x < self.q[i] {
+                        k = i;
+                        break;
+                    }
+                }
+                if self.q[4] < x {
+                    self.q[4] = x;
+                }
+            };
+
+            for i in k..5 {
+                self.n[i] += 1;
+            }
+            for i in 0..5 {
+                self.m[i] += self.dm[i];
+            }
+
+            for i in 1..4 {
+                let d = self.m[i] - self.n[i].to_f64().unwrap();
+                if d >= 1. && self.n[i + 1] - self.n[i] > 1
+                    || d <= -1. && self.n[i - 1] - self.n[i] < -1
+                {
+                    let d = Float::signum(d);
+                    let q_new = self.parabolic(i, d);
+                    if self.q[i - 1] < q_new && q_new < self.q[i + 1] {
+                        self.q[i] = q_new;
+                    } else {
+                        self.q[i] = self.linear(i, d);
+                    }
+                    let delta = d.round() as i64;
+                    debug_assert_eq!(delta.abs(), 1);
+                    self.n[i] += delta;
+                }
+            }
+        }
+
+        pub fn with_samples(mut self, samples: impl IntoIterator<Item = f64>) -> Self {
+            for sample in samples {
+                self.add(sample);
+            }
+
+            self
+        }
+    }
+
+    #[derive(Clone, Copy)]
+    struct FloatOrd(f64);
+
+    impl FloatOrd {
+        fn convert(self) -> u64 {
+            let u: u64 = self.0.to_bits();
+            let bit = 1 << (64 - 1);
+            if u & bit == 0 {
+                u | bit
+            } else {
+                !u
+            }
+        }
+    }
+
+    impl std::cmp::Ord for FloatOrd {
+        fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+            self.convert().cmp(&other.convert())
+        }
+    }
+
+    impl std::cmp::PartialOrd for FloatOrd {
+        fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+            self.convert().partial_cmp(&other.convert())
+        }
+    }
+
+    impl std::cmp::Eq for FloatOrd {}
+    impl std::cmp::PartialEq for FloatOrd {
+        fn eq(&self, other: &Self) -> bool {
+            self.convert().eq(&other.convert())
+        }
+    }
+
+    fn sort_floats(v: &mut [f64]) {
+        let v_: &mut [FloatOrd] = unsafe { std::mem::transmute(v) };
```

Comment:
Unless this is performance critical, we should avoid the `transmute` here and go through something like `sort_by_key` instead.

Reply:
Sounds good, updated in b95f2f5ba705ab799281d460febea5c7ca76eda8

I didn't notice any significant performance impact with the dynamic thresholding enabled, and in any case the author doesn't recommend the thresholding for stable diffusion - I just included it for the sake of completeness.